### PR TITLE
[13.0] Add shopfloor_workstation

### PR DESCRIPTION
**Pull request by TDu**
ref: 1943

### CODE REVIEW

File: shopfloor_mobile/static/wms/src/components/screen.js [hunk 1]
```diff
@@ -59,7 +59,8 @@ Vue.component("Screen", {
                 this.$root.authenticated &&
                 !this.$root.has_profile &&
                 this.$route.name != "profile" &&
-                this.$route.name != "settings"
+                this.$route.name != "settings" &&
+                this.$route.name != "workstation"
```
Comment (simahawk) [inline shopfloor_mobile/static/wms/src/components/screen.js:65]: pls add a TODO that reminds to add a metadata flag to routes (eg: `profile_required`) then we can remove this hardcoded check.

File: shopfloor_mobile/static/wms/src/i18n/i18n.fr.js [hunk 2]
```diff
@@ -28,6 +28,11 @@ const messages_fr = {
                 name: "Profil",
                 title: "Choisissez un profil",
             },
+            workstation: {
+                name: "Sation de travail",
```
Comment (simahawk) [inline shopfloor_mobile/static/wms/src/i18n/i18n.fr.js]:
```suggestion
                name: "Station de travail",
```

File: shopfloor_mobile/static/wms/src/settings/workstation.js [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+        }
+    },
+    methods: {
+        on_scan: function(scanned) {
+            this.odoo.call("setdefault", {barcode: scanned.text}).then(result => {
+                this.workstation_scanned = true;
+                if (result.error) {
+                    console.error(result);
+                    this.scan_message = {
+                        message_type: "error",
```
Comment (simahawk) [inline shopfloor_mobile/static/wms/src/settings/workstation.js]: how do you know that this is the error? You should have it only if you get 404.

BTW We should have something like done here https://github.com/OCA/wms/blob/13.0/shopfloor_mobile/static/wms/src/scenario/mixins.js#L378 but for now is only for scenario.

File: shopfloor_mobile/static/wms/src/settings/workstation.js [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+                        body:
+                            "Request could not be processed, maybe the module shopfloor_workstation is not installed.",
+                    };
+                    return;
+                }
+                this.scan_data = result.data;
+                this.scan_message = result.message;
+                if (this.scan_message.message_type === "info") {
```
Comment (simahawk) [inline shopfloor_mobile/static/wms/src/settings/workstation.js]: you should check if you have the profile or the workstation ID or a specific flag instead.

File: shopfloor_mobile/static/wms/src/settings/workstation.js [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+                    this.$root.trigger(
+                        "profile:selected",
+                        this.scan_data.records[0].profile,
+                        true
+                    );
+                    // TODO: the success message will not be displayed, as we change screen !
```
Comment (simahawk) [inline shopfloor_mobile/static/wms/src/settings/workstation.js:60]: yeah, I did this workaround here https://github.com/OCA/wms/blob/13.0/shopfloor_mobile/static/wms/src/scenario/form.js#L58

File: shopfloor_workstation/services/workstation.py [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+    This allows to scan a workstation bar code on the floor, which
+    will set a default profile on the mobile app and offer the option
+    to change some configuration of the connected user.
+
+    """
+
+    _inherit = "base.shopfloor.service"
+    _name = "shopfloor.workstation"
+    _usage = "workstation"
+    _expose_model = "shopfloor.workstation"
```
Comment (simahawk) [inline shopfloor_workstation/services/workstation.py]: expose_model is not needed if you don't use it

File: shopfloor_workstation/services/workstation.py [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+    _description = __doc__
+
+    def setdefault(self, barcode):
+        """Endpoint that receives a scanned barcode."""
+        ws = self.env["shopfloor.workstation"].search([("barcode", "=", barcode)])
+        if self._set_workstation_as_default(ws):
+            message = {
+                "message_type": "info",
+                "body": "Default workstation set to {}".format(ws.name),
```
Comment (simahawk) [inline shopfloor_workstation/services/workstation.py]: make it translatable

File: shopfloor_workstation/tests/test_shopfloor_workstation.py [hunk 8]
```diff
@@ -0,0 +1,65 @@
+# Copyright 2020 Camptocamp SA (http://www.camptocamp.com)
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from odoo.addons.shopfloor.tests.common import CommonCase
+
+
+class ShopfloorWorkstationCase(CommonCase):
+    def setUp(self):
```
Comment (simahawk) [inline shopfloor_workstation/tests/test_shopfloor_workstation.py]: use setUpClass

**Review by simahawk — CHANGES_REQUESTED** [review]

File: shopfloor_workstation/services/workstation.py [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+            message=message, data={"size": len(ws), "records": self._to_json(ws)}
+        )
+
+    def _set_workstation_as_default(self, workstation):
+        """Apply changes to backend user configuration."""
+        if not workstation:
+            return False
+        if workstation.printing_printer_id:
+            # TODO : should the default action be checked ?
+            self.env.user.printing_printer_id = workstation.printing_printer_id
```
Comment (guewen) [inline shopfloor_workstation/services/workstation.py]: Does the user always has permission to write on this field?

File: shopfloor_mobile/static/wms/src/components/screen.js (hunk 1 above)
Comment (TDu) [reply shopfloor_mobile/static/wms/src/components/screen.js:65]: Added just above

File: shopfloor_workstation/services/workstation.py (hunk 9 above)
Comment (TDu) [reply shopfloor_workstation/services/workstation.py]: Yes he should, it is taken care of there https://github.com/OCA/report-print-send/blob/2160eedf36248d5b7660c5e41a4f03408271668c/base_report_to_printer/models/res_users.py#L28

File: shopfloor_workstation/tests/test_shopfloor_workstation.py (hunk 8 above)
Comment (TDu) [reply shopfloor_workstation/tests/test_shopfloor_workstation.py]: Done

File: shopfloor_workstation/services/workstation.py (hunk 7 above)
Comment (TDu) [reply shopfloor_workstation/services/workstation.py]: Done

File: shopfloor_workstation/services/workstation.py (hunk 6 above)
Comment (TDu) [reply shopfloor_workstation/services/workstation.py]: Removed

File: shopfloor_mobile/static/wms/src/settings/workstation.js (hunk 4 above)
Comment (TDu) [reply shopfloor_mobile/static/wms/src/settings/workstation.js]: I have improved the check

**Comment by TDu** [comment]
@simahawk I updated the PR with some fixes. I had to do a change on `odoo.js` which should be carefully reviewed, thanks.

File: shopfloor_mobile/static/wms/src/services/odoo.js [hunk 10]
```diff
@@ -39,33 +39,29 @@ export class OdooMixin {
...
-                });
+                return {error: handler.call(this, response)};
             }
             return response.json();
         });
     }
-    _error_info(response, json) {
-        return _.extend({}, json, {
-            // strip html
-            description: $(json.description).text(),
```
Comment (simahawk) [inline shopfloor_mobile/static/wms/src/services/odoo.js]: hmm this is required to properly display the error https://github.com/OCA/wms/blob/13.0/shopfloor_mobile/static/wms/src/scenario/mixins.js#L378

File: shopfloor_mobile/static/wms/src/settings/workstation.js [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+        this.odoo = this.$root.getOdoo(odoo_params);
+        if (this.$route.params.identifier) {
+            this.getData(this.$route.params.identifier);
+        }
+    },
+    methods: {
+        on_scan: function(scanned) {
+            this.odoo.call("setdefault", {barcode: scanned.text}).then(result => {
+                this.workstation_scanned = true;
+                if (result.error) {
```
Comment (simahawk) [inline shopfloor_mobile/static/wms/src/settings/workstation.js]: what are you missing here that justifies the change above?

Also, I don't think this piece of code is required.
I would a TODO for something better: when app config is loaded return a mapping of features that are enabled.
This way you could simply check `this.$root.appconfig.features.xyz`

File: shopfloor_workstation/services/workstation.py [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+    """
+
+    _inherit = "base.shopfloor.service"
+    _name = "shopfloor.workstation"
+    _usage = "workstation"
+    _description = __doc__
+
+    def setdefault(self, barcode):
+        """Endpoint that receives a scanned barcode."""
+        ws = self.env["shopfloor.workstation"].search([("barcode", "=", barcode)])
```
Comment (simahawk) [inline shopfloor_workstation/services/workstation.py:26]:
```suggestion
        ws = self.env["shopfloor.workstation"].search([("barcode", "=", barcode)], limit=1)
```

File: shopfloor_workstation/services/workstation.py [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+                "message_type": "info",
+                "body": _("Default workstation set to {}").format(ws.name),
+            }
+        else:
+            message = {
+                "message_type": "error",
+                "body": _("Workstation not found"),
+            }
+        return self._response(
+            message=message, data={"size": len(ws), "records": self._to_json(ws)}
```
Comment (simahawk) [inline shopfloor_workstation/services/workstation.py]: why do you return `records`? Here you should have a single record (the workstation) or nothing.
And you don't need `self._to_json`. Just run `_convert_one_record` or `ws.jsonify($parser, one=True)`

File: shopfloor_workstation/services/workstation.py [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+            }
+        )
+
+    @property
+    def _record_schema(self):
+        return {
+            "id": {"coerce": to_int, "required": True, "type": "integer"},
+            "name": {"type": "string", "nullable": False, "required": True},
+            "barcode": {"type": "string", "nullable": False, "required": True},
+            "profile": {
```
Comment (simahawk) [inline shopfloor_workstation/services/workstation.py:84]: FTR, you could do `self.schemas._schema_dict_of(self.schemas._simple_record())`

File: shopfloor_workstation/views/shopfloor_workstation.xml [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+        <field name="arch" type="xml">
+            <form>
+                <sheet>
+                    <widget
+                        name="web_ribbon"
+                        title="Archived"
+                        bg_color="bg-danger"
+                        attrs="{'invisible': [('active', '=', True)]}"
+                    />
+                    <group>
```
Comment (simahawk) [inline shopfloor_workstation/views/shopfloor_workstation.xml]: name inner groups at least

File: shopfloor_mobile/static/wms/src/components/screen.js [hunk 16]
```diff
@@ -176,6 +179,12 @@ Vue.component("Screen", {
                                 <span>{{ $t('app.profile_configure') }}</span>
                             </v-btn>
                         </v-col>
+                        <v-col class="text-center" cols="12">
+                            <v-btn @click="$router.push({'name': 'workstation'})">
+                                <v-icon>mdi-printer</v-icon>
```
Comment (simahawk) [inline shopfloor_mobile/static/wms/src/components/screen.js]: as the workstation is not just about printers, I'd pick a more generic icon like `cogs` or `account-cog-outline`

**Review by simahawk — CHANGES_REQUESTED** [review]

File: shopfloor_mobile/static/wms/src/services/odoo.js (hunk 10 above)
Comment (TDu) [reply shopfloor_mobile/static/wms/src/services/odoo.js]: But where does the `description` comes from in the response, in the backend ?

Comment (TDu) [reply shopfloor_mobile/static/wms/src/services/odoo.js]: The problem is https://github.com/OCA/wms/pull/145/files#diff-3b7d8d1af55526336a23050aa28b6605a0eb0625d5e129bde2cfd683d8ee42e5L42
crashes with 404 from Odoo (their body contains  html not json data.

File: shopfloor_workstation/services/workstation.py (hunk 12 above)
Comment (TDu) [reply shopfloor_workstation/services/workstation.py:26]: In the latest changes I have added a constraint for the bar code to be unique so not  really necessary

File: shopfloor_workstation/readme/DESCRIPTION.rst [hunk 17]
```diff
@@ -0,0 +1,5 @@
+Shopfloor workstation are useful for a quick configuration of the shopfloor
+mobile application and its connected user.
+
+To synchronize a connected user to a workstation setting, go to
+Settings - Workstation in the mobille application and scan a workstation barcode.
```
Comment (guewen) [inline shopfloor_workstation/readme/DESCRIPTION.rst]:
```suggestion
Settings - Workstation in the mobile application and scan a workstation barcode.
```

File: shopfloor_workstation/services/workstation.py [hunk 18]
```diff
@@ -0,0 +1,100 @@
...
+    _name = "shopfloor.workstation.validator"
+    _usage = "workstation.validator"
+
+    def search(self):
+        return {}
+
+    def get(self):
+        return {
+            "name_fragment": {"type": "string", "nullable": True, "required": False}
+        }
```
Comment (guewen) [inline shopfloor_workstation/services/workstation.py]: Are the search and get endpoints implemented by a super class of the component? Are they used?
I have the impression they do not exist/are not used and could be removed.

Comment (TDu) [reply shopfloor_workstation/services/workstation.py]: Thanks I removed the validator

File: shopfloor_workstation/readme/DESCRIPTION.rst (hunk 17 above)
Comment (TDu) [reply shopfloor_workstation/readme/DESCRIPTION.rst]: Fixed

**Comment by TDu** [comment]
I would like to squash all the messy commits, Let me know when I can do it.

File: shopfloor_mobile/static/wms/src/settings/workstation.js [hunk 19]
```diff
@@ -0,0 +1,69 @@
...
+                        <btn-back/>
+                    </v-col>
+                </v-row>
+            </div>
+        </Screen>
+    `,
+    mounted() {
+        const odoo_params = {usage: this.usage};
+        this.odoo = this.$root.getOdoo(odoo_params);
+        if (this.$route.params.identifier) {
```
Comment (simahawk) [inline shopfloor_mobile/static/wms/src/settings/workstation.js]: who's using this?

File: shopfloor_mobile/static/wms/src/settings/workstation.js [hunk 20]
```diff
@@ -0,0 +1,69 @@
...
+                if (this.scan_data) {
+                    if (this.scan_data.profile) {
+                        this.$root.trigger(
+                            "profile:selected",
+                            this.scan_data.profile,
+                            true
+                        );
+                    }
+                    // TODO: the success message will not be displayed, as we change screen !
+                    this.$root.$router.push({name: "home"});
```
Comment (simahawk) [inline shopfloor_mobile/static/wms/src/settings/workstation.js]: why not getting back to settings page?

Comment (TDu) [reply shopfloor_mobile/static/wms/src/settings/workstation.js]: It was requested like that, but probably make more sense back to settings.
Also they may complain about an extra thumb click

File: shopfloor_mobile/static/wms/src/settings/workstation.js (hunk 19 above)
Comment (TDu) [reply shopfloor_mobile/static/wms/src/settings/workstation.js]: Removed

**Review by simahawk — APPROVED** [review]
squash pls :)

**Review by sebalix — COMMENTED** [review]
Need to add `report-print-send` repo in `oca_dependencies.txt` I think as the depend on `base_report_to_printer`

**Comment by simahawk** [comment]
> Need to add `report-print-send` repo in `oca_dependencies.txt` I think as the depend on `base_report_to_printer`

this is not needed, we have MQT=PIP here.

If you have a pending PR add the pending PR to test-requirements.txt

**Comment by simahawk** [comment]
hopefully fixed here https://github.com/OCA/wms/pull/173

**Pull request closed without merging**